File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/KthSmallestNumberInSortedMatrix.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KthSmallestNumberInSortedMatrix {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int m = in.nextInt();

        int[][] a = new int[n][m];
        for (int i = 0; i < n; i++) {
          for (int j = 0; j < m; j++) {
            a[i][j] = in.nextInt();
          }
        }

        int nq = in.nextInt();
        while (nq-- > 0) {
          int k = in.nextInt();
          System.out.println(kthSmallestNumberInSortedMatix(a, k));
        }
      }
    }
  }

  static class SmallLarge {
    int smallestGreaterThanMid;
    int largestLesserThanOrEqualToMid;
  }

  // O(n log(max - min)) / O(1)
  public static int kthSmallestNumberInSortedMatix(int[][] a, int k) {
    int n = a.length;
    int m = a[0].length;

    int low = a[0][0];
    int high = a[n - 1][m - 1];

    while (low < high) {
      int mid = low + (high - low) / 2;

      SmallLarge range = new SmallLarge();
      int countLesserThanOrEqualToMid = count(a, n, m, mid, range);

      if (countLesserThanOrEqualToMid == k) {
        return range.largestLesserThanOrEqualToMid;
      } else if (countLesserThanOrEqualToMid < k) {
        low = range.smallestGreaterThanMid;
      } else {
        high = range.largestLesserThanOrEqualToMid;
      }
    }

    return low;
  }

  private static int count(int[][] a, int n, int m, int mid, SmallLarge range) {
    int row = n - 1, col = 0;
    int count = 0;

    range.smallestGreaterThanMid = Integer.MAX_VALUE;
    range.largestLesserThanOrEqualToMid = Integer.MIN_VALUE;

    while (row >= 0 && col < m) {
      if (a[row][col] > mid) {
        range.smallestGreaterThanMid =
            Math.min(range.smallestGreaterThanMid, a[row][col]);
        row--;
      } else {
        range.largestLesserThanOrEqualToMid =
            Math.max(range.largestLesserThanOrEqualToMid, a[row][col]);
        count += row + 1;
        col++;
      }
    }

    return count;
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/kth_smallest_number_in_sorted_matrix.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

struct SmallLarge {
  int small;
  int large;

  SmallLarge()
      : small(numeric_limits<int>::max()), large(numeric_limits<int>::min()) {}
};

int process_range(const vector<vector<int>> &a, int n, int m, int mid,
                  SmallLarge &range) {
  int row = n - 1, col = 0;
  int cnt = 0;

  while (row >= 0 && col < m) {
    if (a[row][col] > mid) {
      range.small = min(range.small, a[row][col]);
      row--;
    } else {
      range.large = max(range.large, a[row][col]);
      cnt += row + 1;
      col++;
    }
  }

  return cnt;
}

// O(n log(max - min)) / O(1)
int kth_smallest_in_sorted_matrix(const vector<vector<int>> &a, int k) {
  int n = a.size(), m = a[0].size();
  int low = a[0][0], high = a[n - 1][m - 1];

  while (low < high) {
    int mid = low + (high - low) / 2;

    SmallLarge range;
    int count = process_range(a, n, m, mid, range);

    if (count == k) {
      return range.large;
    } else if (count < k) {
      low = range.small;
    } else {
      high = range.large;
    }
  }

  return low;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m, k, nq;
  cin >> tt;

  while (tt--) {
    cin >> n >> m;

    vector<vector<int>> a(n, vector<int>(m, 0));
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        cin >> a[i][j];
      }
    }

    cin >> nq;
    while (nq--) {
      cin >> k;
      cout << kth_smallest_in_sorted_matrix(a, k) << "\n";
    }
  }

  return 0;
}